src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -716,25 +731,37 @@ protected virtual void Dispose(bool finalize)
         _workQueueTokenSource.Cancel();
     }
 
-    private async ValueTask ProcessWorkQueueAsync(ImmutableSegmentedList<Action> list, CancellationToken cancellationToken)
+    private async ValueTask ProcessEventHandlerWorkQueueAsync(ImmutableSegmentedList<(EventArgs Args, EventHandlerSet HandlerSet)> list, CancellationToken cancellationToken)
     {
-        // Hop over to the right scheduler to execute all this work.
-        await Task.Factory.StartNew(() =>
+        // Currently, this method maintains ordering of events to their callbacks. If that were no longer necessary,
+        // further performance optimizations of this method could be considered, such as grouping together 
+        // callbacks by eventargs/registries or parallelizing callbacks.
+
+        // ABWQ callbacks occur on threadpool threads, so we can process the handlers immediately that don't require the main thread.
+        ProcessWorkQueueHelper(list, shouldRaise: static options => !options.RequiresMainThread, cancellationToken);
+
+        // Verify there are handlers which require the main thread before performing the thread switch.
+        if (list.Any(static x => x.HandlerSet.HasMatchingOptions(isMatch: static options => options.RequiresMainThread)))
```

Comment:
Did you consider having us kick off the work to the main thread prior to doing the background thread ones, and then awaiting the task once we're done? At least that way there's a bit more happening in parallel and the change is fairly trivial.

Reply:
Concurrent handler invocation is definitely something we should consider and worthy of it's own effort as there are a lot of possibilities there with a lot of potential snags.